### ISSUE TITLE
Running benchmarks on TruffleRuby Native

### ISSUE THREAD

**Issue by eregon** [kostya/benchmarks#246]
Could you run the benchmarks on TruffleRuby Native? (which is the default mode, no `--jvm` flag)

I see only TruffleRuby JVM results in the README, and given the benchmarks run basically without warmup (it's just timing the single benchmark iteration, so any warmup is counted against), it would be interesting to run in native mode.

**Comment by eregon** [comment]
BTW, what does this mean in the README? cc @nuald

> JIT warming up is applied when necessary, and the actual measurements are taken only within the specified boundaries.

As far as I can see the benchmark is run only once, without warmup, and yes code loading seems not accounted for but that's all.
That means any JITing happens during the measurement and a good part might run in the interpreter, so it would be good to clarify that in the README.

**Comment by nuald** [comment]
TruffleRuby native has been crashing on the previous environment, however the current environment could handle it better, I'll give it one more chance.

As for the JIT warming up, I guess you've referred to the Ruby code as all actual JIT compiler (e.g. Java, C#, Julia) tests have the warming up code. Ruby's JIT is still experimental and not used in the tests. TruffleRuby based on JVM is another story, and could gain a little from the warming up.

**Comment by eregon** [comment]
Thanks, if it crashes please file an issue report, that would of course be a bug.

Regarding JIT warmup I'm thinking about TruffleRuby Native, TruffleRuby JVM, JRuby, Java, etc's JITs, which all must operate while the benchmark is running and measured and so the benchmark numbers ends up measuring how fast warmup is.
That can be fine as a way to measure, just the README should make that clear it doesn't do anything about JIT warmup then (just avoids measuring startup/code loading).

**Comment by eregon** [comment]
I see what you mean, Java files have explicit warmup:
https://github.com/kostya/benchmarks/blob/c9e1ca7c225d22762388492201b07c21450e2ad4/brainfuck2/bf.java#L131-L133
but Ruby files don't:
https://github.com/kostya/benchmarks/blob/c9e1ca7c225d22762388492201b07c21450e2ad4/brainfuck2/bf.rb#L94

That's essentially not fair for Ruby implementations having a JIT, when compared to Java and other languages which have explicit warmup.

Doing explicit warmup correctly is also hard BTW, the JIT'ed code might be thrown away since the actual benchmark might do something quite different (e.g., take a new branch never taken in the warmup code). And the other way around, warmup code could even make the benchmark slower by compiling things (e.g., branches) which might not happen in the actual benchmark.
I think a good way to improve on that is to run the full benchmark once as warmup, and then once as mesured. Then it's the same workload and the JIT has a chance to optimize the actual code for the benchmark.
Another way is to simply acknowledge that measurements include warmup, which is probably fine if the benchmarks run long enough, but some seem quite short here.

**Comment by nuald** [comment]
I agree that the Ruby code need the warmup section, and it'll be addressed.

TruffleRuby Native crashes have been related to the memory issues, and I'm not even sure that it's the fault of GraalVM, but could be rather some Docker problems. Regardless of that, I've tried the tests again and they work, so the results will be added in the near future.

As for the JIT, please let me note the following:
- measurements do not include warmups - the code performance is being measured only between the `notify` calls;
- I agree that the current warmup is not ideal, however the proper sampling for JIT may require not one, but tens and hundreds samples of the same code running again and again. Right now, we eliminate delays for the class loading, bytecode lookups and initial compilation; sample-based profiling and optimization is out of scope (that's an area of the fine tuning as JIT compilers may have the unfair advantage here as other programming language tests could have the proper micro-optimized code too).
 - In the real world people could have the unpleasant surprises if they would trust the numbers of fine-tuned JIT-friendly tests only. Here we have the middle ground - the code is being loaded and compiled by JIT, but not so optimized yet. Surely, it would be best if we have the cold started and fine tuned versions, but that would make the tests forever to run. As I don't have any server farm for that, but rather only local workstation, I can't do it, sorry.

**Comment by eregon** [comment]
> measurements do not include warmups

Measurements do not include the time to run warmup *code*, but they seem likely to include warmup in term of including time to just-in-compile things.
Since the warmup code is different than the benchmark, it sounds likely the benchmark will need new compilations, and until then it might run partly in the interpreter.
The warmup code might also simply run for a too short time for the JIT to have time to compile anything interesting.

It's a hard problem, no doubt.

> Right now, we eliminate delays for the class loading, bytecode lookups and initial compilation

Agreed, except initial compilation might not happen during execution of the warmup code, and there can be compilation later too.

> Here we have the middle ground - the code is being loaded and compiled by JIT, but not so optimized yet.

That's the part I'm not confident about, the relevant code might not be JITed or might deoptimize just when starting the actual benchmark.

That's why I would advice to either:
* not do anything for warmup since these are small programs and they should compile fast. That would still not count startup/class loading by using the `notify` as they are. And then just mention warmup is part of the measurement, and startup is excluded in the README.
* try to ensure there is no JIT compilation during the benchmark (e.g., by automatically checking compilation logs) but that likely needs warmup code more similar to the benchmark code, running each benchmark many times, etc, a whole different and far more complex setup.

**Comment by nuald** [comment]
Agreed, I've tried different combinations already and looks like removing the warmup code would be the best approach. For example, for TruffleRuby Native bf2 test I got (with Linux kernel is in the performance mode):

- No warmup: 31s
- Warmup with the non-relevant code (as we have for other tests): 38s
- Warmup with the real code (1 iteration): 51s
- Warmup with the real code (10 iterations): 42s

Other JITs have different results (usually few iterations give better results), but for all of them "no warmup" give close to the best results. I guess the slower results after warmup could be caused by the memory consumption and Docker environment, however I'm not completely sure about it. Moreover I doubt it's worth researching as making JIT happy is the fine tuning and not always happening in the real world.

Updating other tests may take a while, and for some of them warmup could be required though, but I'll analyze it per-test.

**Comment by eregon** [comment]
Agreed, I think not running separate warmup here is the best and the simplest for those benchmarks.

**Comment by nuald** [comment]
Unfortunately, just dropping JIT warmup is not always a good idea as other factors are kicking in like lazy initialization (for example, Nd4j lazy initialization takes 0.6s at the same time as the actual test takes 0.2s). As the middle ground, I've decided to just run tests in a loop, please see #248 For usual tests (where no lazy functionality is used) it means that each test would run twice (unless the second iteration is faster than 1st).

For that ticket I'll just add TruffleRuby native results. If you would like to discuss any other improvements in the testing framework, please use the aforementioned ticket.

**Comment by nuald** [comment]
Addressed in #249
